File: frontend/src/services/moneyRequests.ts
/**
 * Money Request service for SoftBankCashWire frontend
 */
import apiClient, { handleApiError } from './api'
import { MoneyRequest, RequestStatus, Transaction } from '../types'

export interface CreateMoneyRequestRequest {
  recipient_id: string
  amount: string
  note?: string
  expires_in_days?: number
}

export interface CreateMoneyRequestResult {
  success: boolean
  request: MoneyRequest
  expires_in_hours: number
}

export interface RespondToRequestRequest {
  approved: boolean
}

export interface RespondToRequestResult {
  success: boolean
  approved: boolean
  request: MoneyRequest
  transaction?: Transaction
  sender_balance?: string
  recipient_balance?: string
  message?: string
}

export interface CancelRequestResult {
  success: boolean
  request: MoneyRequest
  message: string
}

export interface MoneyRequestList {
  requests: MoneyRequest[]
  pagination?: {
    total: number
    limit: number
    offset: number
    has_more: boolean
  }
  count?: number
}

export interface MoneyRequestStatistics {
  period_days: number
  sent_requests: {
    total: number
    approved: number
    declined: number
    pending: number
    expired: number
    approval_rate: number
    total_amount_approved: string
  }
  received_requests: {
    total: number
    approved: number
    declined: number
    pending: number
    expired: number
    approval_rate: number
    total_amount_approved: string
  }
}

export interface RequestValidationResult {
  valid: boolean
  errors: Array<{
    code: string
    message: string
  }>
  warnings: Array<{
    code: string
    message: string
  }>
}

export interface ExpiringRequestsResult {
  requests: MoneyRequest[]
  count: number
  hours_threshold: number
}

class MoneyRequestsService {
  /**
   * Create a new money request
   */
  static async createMoneyRequest(request: CreateMoneyRequestRequest): Promise<CreateMoneyRequestResult> {
    try {
      const response = await apiClient.post('/money-requests/create', request)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Respond to a money request (approve or decline)
   */
  static async respondToRequest(requestId: string, approved: boolean): Promise<RespondToRequestResult> {
    try {
      const response = await apiClient.post(`/money-requests/${requestId}/respond`, {
        approved
      })
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Cancel a money request
   */
  static async cancelRequest(requestId: string): Promise<CancelRequestResult> {
    try {
      const response = await apiClient.post(`/money-requests/${requestId}/cancel`)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get money request details by ID
   */
  static async getRequest(requestId: string): Promise<{ request: MoneyRequest }> {
    try {
      const response = await apiClient.get(`/money-requests/${requestId}`)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get pending money requests (as recipient)
   */
  static async getPendingRequests(): Promise<MoneyRequestList> {
    try {
      const response = await apiClient.get('/money-requests/pending')
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get sent money requests
   */
  static async getSentRequests(
    status?: RequestStatus,
    limit: number = 50,
    offset: number = 0
  ): Promise<MoneyRequestList> {
    try {
      const params = new URLSearchParams()
      if (status) params.append('status', status)
      params.append('limit', limit.toString())
      params.append('offset', offset.toString())

      const response = await apiClient.get(`/money-requests/sent?${params.toString()}`)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get received money requests
   */
  static async getReceivedRequests(
    status?: RequestStatus,
    limit: number = 50,
    offset: number = 0
  ): Promise<MoneyRequestList> {
    try {
      const params = new URLSearchParams()
      if (status) params.append('status', status)
      params.append('limit', limit.toString())
      params.append('offset', offset.toString())

      const response = await apiClient.get(`/money-requests/received?${params.toString()}`)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get money request statistics
   */
  static async getRequestStatistics(days: number = 30): Promise<MoneyRequestStatistics> {
    try {
      const response = await apiClient.get(`/money-requests/statistics?days=${days}`)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Validate money request creation
   */
  static async validateRequestCreation(
    recipientId: string,
    amount: string
  ): Promise<RequestValidationResult> {
    try {
      const response = await apiClient.post('/money-requests/validate', {
        recipient_id: recipientId,
        amount: amount
      })
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get expiring requests
   */
  static async getExpiringRequests(hours: number = 24): Promise<ExpiringRequestsResult> {
    try {
      const response = await apiClient.get(`/money-requests/expiring?hours=${hours}`)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Format currency amount for display
   */
  static formatCurrency(amount: string, currency: string = 'GBP'): string {
    const numAmount = parseFloat(amount)
    
    if (currency === 'GBP') {
      return new Intl.NumberFormat('en-GB', {
        style: 'currency',
        currency: 'GBP',
      }).format(numAmount)
    }
    
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: currency,
    }).format(numAmount)
  }

  /**
   * Get request status color for UI
   */
  static getRequestStatusColor(status: RequestStatus): 'success' | 'warning' | 'error' | 'info' {
    switch (status) {
      case RequestStatus.APPROVED:
        return 'success'
      case RequestStatus.DECLINED:
        return 'error'
      case RequestStatus.EXPIRED:
        return 'error'
      case RequestStatus.PENDING:
        return 'warning'
      default:
        return 'info'
    }
  }

  /**
   * Get request status display name
   */
  static getRequestStatusDisplayName(status: RequestStatus): string {
    switch (status) {
      case RequestStatus.PENDING:
        return 'Pending'
      case RequestStatus.APPROVED:
        return 'Approved'
      case RequestStatus.DECLINED:
        return 'Declined'
      case RequestStatus.EXPIRED:
        return 'Expired'
      default:
        return 'Unknown'
    }
  }

  /**
   * Check if request is expired
   */
  static isRequestExpired(request: MoneyRequest): boolean {
    if (!request.expires_at) return false
    return new Date(request.expires_at) < new Date()
  }

  /**
   * Check if request is expiring soon
   */
  static isRequestExpiringSoon(request: MoneyRequest, hoursThreshold: number = 24): boolean {
    if (!request.expires_at || request.status !== RequestStatus.PENDING) return false
    
    const expiryTime = new Date(request.expires_at).getTime()
    const now = new Date().getTime()
    const hoursUntilExpiry = (expiryTime - now) / (1000 * 60 * 60)
    
    return hoursUntilExpiry <= hoursThreshold && hoursUntilExpiry > 0
  }

  /**
   * Get time until expiry
   */
  static getTimeUntilExpiry(request: MoneyRequest): {
    expired: boolean
    days: number
    hours: number
    minutes: number
  } {
    if (!request.expires_at) {
      return { expired: false, days: 0, hours: 0, minutes: 0 }
    }

    const expiryTime = new Date(request.expires_at).getTime()
    const now = new Date().getTime()
    const timeDiff = expiryTime - now

    if (timeDiff <= 0) {
      return { expired: true, days: 0, hours: 0, minutes: 0 }
    }

    const days = Math.floor(timeDiff / (1000 * 60 * 60 * 24))
    const hours = Math.floor((timeDiff % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60))
    const minutes = Math.floor((timeDiff % (1000 * 60 * 60)) / (1000 * 60))

    return { expired: false, days, hours, minutes }
  }

  /**
   * Format time until expiry for display
   */
  static formatTimeUntilExpiry(request: MoneyRequest): string {
    const timeInfo = this.getTimeUntilExpiry(request)
    
    if (timeInfo.expired) {
      return 'Expired'
    }

    if (timeInfo.days > 0) {
      return `${timeInfo.days} day${timeInfo.days > 1 ? 's' : ''} remaining`
    }

    if (timeInfo.hours > 0) {
      return `${timeInfo.hours} hour${timeInfo.hours > 1 ? 's' : ''} remaining`
    }

    if (timeInfo.minutes > 0) {
      return `${timeInfo.minutes} minute${timeInfo.minutes > 1 ? 's' : ''} remaining`
    }

    return 'Expiring soon'
  }

  /**
   * Validate request amount
   */
  static validateAmount(amount: string): { valid: boolean; error?: string } {
    if (!amount || amount.trim() === '') {
      return { valid: false, error: 'Amount is required' }
    }

    const numAmount = parseFloat(amount)
    
    if (isNaN(numAmount)) {
      return { valid: false, error: 'Amount must be a valid number' }
    }

    if (numAmount <= 0) {
      return { valid: false, error: 'Amount must be positive' }
    }

    if (numAmount > 10000) {
      return { valid: false, error: 'Amount cannot exceed £10,000' }
    }

    // Check for reasonable decimal places
    const decimalPlaces = (amount.split('.')[1] || '').length
    if (decimalPlaces > 2) {
      return { valid: false, error: 'Amount cannot have more than 2 decimal places' }
    }

    return { valid: true }
  }

  /**
   * Validate request note
   */
  static validateNote(note?: string): { valid: boolean; error?: string } {
    if (!note) {
      return { valid: true } // Note is optional
    }

    if (note.length > 500) {
      return { valid: false, error: 'Note cannot exceed 500 characters' }
    }

    return { valid: true }
  }

  /**
   * Validate expiry days
   */
  static validateExpiryDays(days?: number): { valid: boolean; error?: string } {
    if (days === undefined || days === null) {
      return { valid: true } // Optional, will use default
    }

    if (!Number.isInteger(days) || days < 1 || days > 30) {
      return { valid: false, error: 'Expiry days must be between 1 and 30' }
    }

    return { valid: true }
  }

  /**
   * Group requests by status
   */
  static groupRequestsByStatus(requests: MoneyRequest[]): Record<RequestStatus, MoneyRequest[]> {
    const grouped: Record<RequestStatus, MoneyRequest[]> = {
      [RequestStatus.PENDING]: [],
      [RequestStatus.APPROVED]: [],
      [RequestStatus.DECLINED]: [],
      [RequestStatus.EXPIRED]: [],
    }
    
    requests.forEach(request => {
      grouped[request.status].push(request)
    })
    
    return grouped
  }

  /**
   * Sort requests by various criteria
   */
  static sortRequests(
    requests: MoneyRequest[], 
    sortBy: 'date' | 'amount' | 'status' | 'expiry',
    sortOrder: 'asc' | 'desc' = 'desc'
  ): MoneyRequest[] {
    const sorted = [...requests].sort((a, b) => {
      let comparison = 0

      switch (sortBy) {
        case 'date':
          comparison = new Date(a.created_at).getTime() - new Date(b.created_at).getTime()
          break
        case 'amount':
          comparison = parseFloat(a.amount) - parseFloat(b.amount)
          break
        case 'status':
          comparison = a.status.localeCompare(b.status)
          break
        case 'expiry':
          const aExpiry = a.expires_at ? new Date(a.expires_at).getTime() : 0
          const bExpiry = b.expires_at ? new Date(b.expires_at).getTime() : 0
          comparison = aExpiry - bExpiry
          break
      }

      return sortOrder === 'asc' ? comparison : -comparison
    })

    return sorted
  }

  /**
   * Filter requests by text search
   */
  static searchRequests(requests: MoneyRequest[], searchText: string): MoneyRequest[] {
    if (!searchText.trim()) {
      return requests
    }

    const searchLower = searchText.toLowerCase()
    
    return requests.filter(request => {
      return (
        request.note?.toLowerCase().includes(searchLower) ||
        request.requester_name?.toLowerCase().includes(searchLower) ||
        request.recipient_name?.toLowerCase().includes(searchLower) ||
        request.amount.includes(searchText) ||
        request.status.toLowerCase().includes(searchLower)
      )
    })
  }

  /**
   * Get request summary for display
   */
  static getRequestSummary(requests: MoneyRequest[], currentUserId: string): {
    totalSent: number
    totalReceived: number
    pendingSent: number
    pendingReceived: number
    approvedSent: number
    approvedReceived: number
  } {
    let totalSent = 0
    let totalReceived = 0
    let pendingSent = 0
    let pendingReceived = 0
    let approvedSent = 0
    let approvedReceived = 0

    requests.forEach(request => {
      const amount = parseFloat(request.amount)
      
      if (request.requester_id === currentUserId) {
        // Sent request
        totalSent += amount
        if (request.status === RequestStatus.PENDING) {
          pendingSent += amount
        } else if (request.status === RequestStatus.APPROVED) {
          approvedSent += amount
        }
      } else {
        // Received request
        totalReceived += amount
        if (request.status === RequestStatus.PENDING) {
          pendingReceived += amount
        } else if (request.status === RequestStatus.APPROVED) {
          approvedReceived += amount
        }
      }
    })

    return {
      totalSent,
      totalReceived,
      pendingSent,
      pendingReceived,
      approvedSent,
      approvedReceived,
    }
  }

  /**
   * Check if user can respond to request
   */
  static canRespondToRequest(request: MoneyRequest, currentUserId: string): boolean {
    return (
      request.recipient_id === currentUserId &&
      request.status === RequestStatus.PENDING &&
      !this.isRequestExpired(request)
    )
  }

  /**
   * Check if user can cancel request
   */
  static canCancelRequest(request: MoneyRequest, currentUserId: string): boolean {
    return (
      request.requester_id === currentUserId &&
      request.status === RequestStatus.PENDING
    )
  }
}

export const moneyRequestsService = MoneyRequestsService
export default MoneyRequestsService

File: backend/api/money_requests.py
"""
Money Request API endpoints for SoftBankCashWire
"""
from flask import Blueprint, request, jsonify, g
from decimal import Decimal, InvalidOperation
from services.money_request_service import MoneyRequestService
from middleware.auth_middleware import auth_required, get_client_info, validate_request_data
from models import db, RequestStatus

money_requests_bp = Blueprint('money_requests', __name__)

@money_requests_bp.route('/create', methods=['POST'])
@auth_required
@validate_request_data(['recipient_id', 'amount'])
def create_money_request():
    """
    Create a new money request
    
    Expected JSON:
        {
            "recipient_id": "user-id",
            "amount": "50.00",
            "note": "Lunch payment",  // Optional
            "expires_in_days": 7  // Optional, default 7, max 30
        }
    
    Returns:
        JSON with request creation result
    """
    try:
        requester_id = g.current_user_id
        data = request.get_json()
        
        recipient_id = data['recipient_id']
        
        # Parse and validate amount
        try:
            amount = Decimal(str(data['amount']))
        except (InvalidOperation, ValueError):
            return jsonify({
                'error': {
                    'code': 'INVALID_AMOUNT',
                    'message': 'Amount must be a valid number'
                }
            }), 400
        
        note = data.get('note')
        expires_in_days = data.get('expires_in_days')
        
        # Validate note length
        if note and len(note) > 500:
            return jsonify({
                'error': {
                    'code': 'NOTE_TOO_LONG',
                    'message': 'Note cannot exceed 500 characters'
                }
            }), 400
        
        # Validate expires_in_days
        if expires_in_days is not None:
            try:
                expires_in_days = int(expires_in_days)
            except (ValueError, TypeError):
                return jsonify({
                    'error': {
                        'code': 'INVALID_EXPIRY',
                        'message': 'Expiry days must be a valid number'
                    }
                }), 400
        
        ip_address, user_agent = get_client_info()
        
        # Create money request
        result = MoneyRequestService.create_money_request(
            requester_id=requester_id,
            recipient_id=recipient_id,
            amount=amount,
            note=note,
            expires_in_days=expires_in_days,
            ip_address=ip_address,
            user_agent=user_agent
        )
        
        return jsonify(result), 201
        
    except ValueError as e:
        return jsonify({
            'error': {
                'code': 'REQUEST_CREATION_FAILED',
                'message': str(e)
            }
        }), 400
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'CREATE_REQUEST_ERROR',
                'message': f'Failed to create money request: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/<request_id>/respond', methods=['POST'])
@auth_required
@validate_request_data(['approved'])
def respond_to_request(request_id):
    """
    Respond to a money request (approve or decline)
    
    Expected JSON:
        {
            "approved": true  // true to approve, false to decline
        }
    
    Returns:
        JSON with response result
    """
    try:
        user_id = g.current_user_id
        data = request.get_json()
        
        approved = data['approved']
        
        if not isinstance(approved, bool):
            return jsonify({
                'error': {
                    'code': 'INVALID_RESPONSE',
                    'message': 'Approved must be true or false'
                }
            }), 400
        
        ip_address, user_agent = get_client_info()
        
        # Respond to request
        result = MoneyRequestService.respond_to_request(
            request_id=request_id,
            user_id=user_id,
            approved=approved,
            ip_address=ip_address,
            user_agent=user_agent
        )
        
        return jsonify(result), 200
        
    except ValueError as e:
        return jsonify({
            'error': {
                'code': 'RESPONSE_FAILED',
                'message': str(e)
            }
        }), 400
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'RESPOND_ERROR',
                'message': f'Failed to respond to request: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/<request_id>/cancel', methods=['POST'])
@auth_required
def cancel_request(request_id):
    """
    Cancel a money request (only by requester)
    
    Returns:
        JSON with cancellation result
    """
    try:
        user_id = g.current_user_id
        ip_address, user_agent = get_client_info()
        
        result = MoneyRequestService.cancel_request(
            request_id=request_id,
            user_id=user_id,
            ip_address=ip_address,
            user_agent=user_agent
        )
        
        return jsonify(result), 200
        
    except ValueError as e:
        return jsonify({
            'error': {
                'code': 'CANCEL_FAILED',
                'message': str(e)
            }
        }), 400
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'CANCEL_ERROR',
                'message': f'Failed to cancel request: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/<request_id>', methods=['GET'])
@auth_required
def get_request(request_id):
    """
    Get money request details by ID
    
    Returns:
        JSON with request details
    """
    try:
        user_id = g.current_user_id
        
        money_request = MoneyRequestService.get_request_by_id(request_id, user_id)
        
        if not money_request:
            return jsonify({
                'error': {
                    'code': 'REQUEST_NOT_FOUND',
                    'message': 'Money request not found or access denied'
                }
            }), 404
        
        return jsonify({
            'request': money_request.to_dict(include_names=True)
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'GET_REQUEST_ERROR',
                'message': f'Failed to get request: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/pending', methods=['GET'])
@auth_required
def get_pending_requests():
    """
    Get pending money requests for authenticated user (as recipient)
    
    Returns:
        JSON with pending requests
    """
    try:
        user_id = g.current_user_id
        
        pending_requests = MoneyRequestService.get_pending_requests_for_user(user_id)
        
        return jsonify({
            'requests': [req.to_dict(include_names=True) for req in pending_requests],
            'count': len(pending_requests)
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'PENDING_REQUESTS_ERROR',
                'message': f'Failed to get pending requests: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/sent', methods=['GET'])
@auth_required
def get_sent_requests():
    """
    Get money requests sent by authenticated user
    
    Query Parameters:
        - status: Filter by status (PENDING, APPROVED, DECLINED, EXPIRED)
        - limit: Number of requests to return (default: 50, max: 100)
        - offset: Number of requests to skip (default: 0)
    
    Returns:
        JSON with sent requests and pagination
    """
    try:
        user_id = g.current_user_id
        
        # Parse query parameters
        status = None
        if request.args.get('status'):
            try:
                status = RequestStatus(request.args.get('status'))
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_STATUS',
                        'message': 'Invalid status value'
                    }
                }), 400
        
        limit = 50
        if request.args.get('limit'):
            try:
                limit = int(request.args.get('limit'))
                if limit < 1 or limit > 100:
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_LIMIT',
                        'message': 'Limit must be between 1 and 100'
                    }
                }), 400
        
        offset = 0
        if request.args.get('offset'):
            try:
                offset = int(request.args.get('offset'))
                if offset < 0:
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_OFFSET',
                        'message': 'Offset must be non-negative'
                    }
                }), 400
        
        # Get sent requests
        result = MoneyRequestService.get_sent_requests(
            user_id=user_id,
            status=status,
            limit=limit,
            offset=offset
        )
        
        return jsonify(result), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'SENT_REQUESTS_ERROR',
                'message': f'Failed to get sent requests: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/received', methods=['GET'])
@auth_required
def get_received_requests():
    """
    Get money requests received by authenticated user
    
    Query Parameters:
        - status: Filter by status (PENDING, APPROVED, DECLINED, EXPIRED)
        - limit: Number of requests to return (default: 50, max: 100)
        - offset: Number of requests to skip (default: 0)
    
    Returns:
        JSON with received requests and pagination
    """
    try:
        user_id = g.current_user_id
        
        # Parse query parameters (same as sent requests)
        status = None
        if request.args.get('status'):
            try:
                status = RequestStatus(request.args.get('status'))
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_STATUS',
                        'message': 'Invalid status value'
                    }
                }), 400
        
        limit = 50
        if request.args.get('limit'):
            try:
                limit = int(request.args.get('limit'))
                if limit < 1 or limit > 100:
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_LIMIT',
                        'message': 'Limit must be between 1 and 100'
                    }
                }), 400
        
        offset = 0
        if request.args.get('offset'):
            try:
                offset = int(request.args.get('offset'))
                if offset < 0:
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_OFFSET',
                        'message': 'Offset must be non-negative'
                    }
                }), 400
        
        # Get received requests
        result = MoneyRequestService.get_received_requests(
            user_id=user_id,
            status=status,
            limit=limit,
            offset=offset
        )
        
        return jsonify(result), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'RECEIVED_REQUESTS_ERROR',
                'message': f'Failed to get received requests: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/statistics', methods=['GET'])
@auth_required
def get_request_statistics():
    """
    Get money request statistics for authenticated user
    
    Query Parameters:
        - days: Number of days to analyze (default: 30, max: 365)
    
    Returns:
        JSON with request statistics
    """
    try:
        user_id = g.current_user_id
        
        # Parse days parameter
        days = 30
        if request.args.get('days'):
            try:
                days = int(request.args.get('days'))
                if days < 1 or days > 365:
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_DAYS',
                        'message': 'Days must be between 1 and 365'
                    }
                }), 400
        
        # Get statistics
        statistics = MoneyRequestService.get_request_statistics(user_id, days)
        
        return jsonify(statistics), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'STATISTICS_ERROR',
                'message': f'Failed to get request statistics: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/validate', methods=['POST'])
@auth_required
@validate_request_data(['recipient_id', 'amount'])
def validate_request_creation():
    """
    Validate money request creation without creating it
    
    Expected JSON:
        {
            "recipient_id": "user-id",
            "amount": "50.00"
        }
    
    Returns:
        JSON with validation results
    """
    try:
        requester_id = g.current_user_id
        data = request.get_json()
        
        recipient_id = data['recipient_id']
        
        # Parse amount
        try:
            amount = Decimal(str(data['amount']))
        except (InvalidOperation, ValueError):
            return jsonify({
                'error': {
                    'code': 'INVALID_AMOUNT',
                    'message': 'Amount must be a valid number'
                }
            }), 400
        
        # Validate request
        validation = MoneyRequestService.validate_request_creation(
            requester_id=requester_id,
            recipient_id=recipient_id,
            amount=amount
        )
        
        return jsonify(validation), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'VALIDATION_ERROR',
                'message': f'Failed to validate request: {str(e)}'
            }
        }), 500

@money_requests_bp.route('/expiring', methods=['GET'])
@auth_required
def get_expiring_requests():
    """
    Get requests that are expiring soon (admin/system endpoint)
    
    Query Parameters:
        - hours: Hours until expiration (default: 24)
    
    Returns:
        JSON with expiring requests
    """
    try:
        # Parse hours parameter
        hours = 24
        if request.args.get('hours'):
            try:
                hours = int(request.args.get('hours'))
                if hours < 1 or hours > 168:  # Max 1 week
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_HOURS',
                        'message': 'Hours must be between 1 and 168'
                    }
                }), 400
        
        # Get expiring requests
        expiring_requests = MoneyRequestService.get_expiring_requests(hours)
        
        return jsonify({
            'requests': [req.to_dict(include_names=True) for req in expiring_requests],
            'count': len(expiring_requests),
            'hours_threshold': hours
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'EXPIRING_REQUESTS_ERROR',
                'message': f'Failed to get expiring requests: {str(e)}'
            }
        }), 500

# Error handlers for money requests blueprint
@money_requests_bp.errorhandler(400)
def bad_request(error):
    """Handle bad request errors"""
    return jsonify({
        'error': {
            'code': 'BAD_REQUEST',
            'message': 'Invalid request format'
        }
    }), 400

@money_requests_bp.errorhandler(401)
def unauthorized(error):
    """Handle unauthorized errors"""
    return jsonify({
        'error': {
            'code': 'UNAUTHORIZED',
            'message': 'Authentication required'
        }
    }), 401

@money_requests_bp.errorhandler(404)
def not_found(error):
    """Handle not found errors"""
    return jsonify({
        'error': {
            'code': 'NOT_FOUND',
            'message': 'Resource not found'
        }
    }), 404

@money_requests_bp.errorhandler(500)
def internal_error(error):
    """Handle internal server errors"""
    db.session.rollback()
    return jsonify({
        'error': {
            'code': 'INTERNAL_ERROR',
            'message': 'Internal server error'
        }
    }), 500